ui/ducks/send/send.js

```diff
@@ -0,0 +1,1238 @@
+    const {
+      gas: { basicEstimateStatus, basicEstimates },
+    } = thunkApi.getState();
+    // Default gasPrice to 1 gwei if all estimation fails
+    const gasPrice =
+      basicEstimateStatus === BASIC_ESTIMATE_STATES.READY
+        ? getGasPriceInHexWei(basicEstimates.average)
+        : '0x1';
+    // Set a basic gasLimit in the event that other estimation fails
+    let gasLimit =
+      asset.type === ASSET_TYPES.TOKEN
+        ? GAS_LIMITS.BASE_TOKEN_ESTIMATE
+        : GAS_LIMITS.SIMPLE;
+    if (
+      basicEstimateStatus === BASIC_ESTIMATE_STATES.READY &&
+      stage !== SEND_STAGES.EDIT
+    ) {
+      // Run our estimateGasLimit logic to get a more accurate estimation of
+      // required gas. If this value isn't nullish, set it as the new gasLimit
+      const estimatedGasLimit = await estimateGasLimitForSend({
+        gasPrice: getGasPriceInHexWei(basicEstimates.average),
+        blockGasLimit: metamask.blockGasLimit,
+        selectedAddress: fromAddress,
+        sendToken: asset.details,
+        to: recipient.address.toLowerCase(),
+        value: amount.value,
+        data: draftTransaction.userInputHexData,
+      });
+      gasLimit ??= estimatedGasLimit;
+    }
+    // We have to keep the customGas setting in sync with the gasLimit when
+    // changing it due to side effects of user actions.
+    await thunkApi.dispatch(setCustomGasLimit(gasLimit));
+    // We must determine the balance of the asset that the transaction will be
+    // sending. This is done by referencing the native balance on the account
+    // for native assets, and calling the balanceOf method on the ERC20
+    // contract for token sends.
+    let { balance } = account;
+    if (asset.type === ASSET_TYPES.TOKEN) {
+      if (asset.details === null) {
+        // If we're sending a token but details have not been provided we must
+        // abort and set the send slice into invalid status.
+        throw new Error(
+          'Send slice initialized as token send without token details',
+        );
+      }
+      balance = await getERC20Balance(asset.details, fromAddress);
+    }
+    return {
+      address: fromAddress,
+      nativeBalance: account.balance,
+      assetBalance: balance,
+      chainId: getCurrentChainId(state),
+      tokens: getTokens(state),
+      gasPrice,
+      gasLimit,
+      gasTotal: addHexPrefix(calcGasTotal(gasLimit, gasPrice)),
+    };
+  },
+);
+
+export const initialState = {
+  // which stage of the send flow is the user on
+  stage: SEND_STAGES.UNINITIALIZED,
+  // status of the send slice, either VALID or INVALID
+  status: SEND_STATUSES.VALID,
+  account: {
+    // from account address, defaults to selected account. will be the account
+    // the original transaction was sent from in the case of the EDIT stage
+    address: null,
+    // balance of the from account
+    balance: '0x0',
+  },
+  gas: {
+    // which gas estimation mode is being used
+    mode: 'BASIC',
```

Comment:
Isn't this part of the background state, in the preferences controller? Or maybe this is referring to something different? Not sure exactly what this is for.

Reply:
used to control which estimation UI to show on the send page. Not in love with this and open to changing or at the very least better documenting it.

Reply:
The estimation UI to show is determined by a few different things:

* The `advancedInlineGas` feature flag, set by the "Advanced gas controls" setting, should set it to always use Advanced
* On networks not supported by our gas estimation API, we will show advanced
* If the gas estimation API is down, we will shown advanced
* Otherwise, we always show basic

I'm unsure what role this state plays. I tried looking for cases where this state was changed to basic or advanced, and it looks like it's _never_ set to basic. And the one place it's set to Advanced is during an `onSubmit` function, where it has no obvious purpose (I suspect it was meant to prevent changing the gas after clicking submit, but before the next page had loaded? Who knows).

Reply:
It might make sense to move the logic I just described into the slice. But it looks like at this time, that hasn't been done yet - it appears to still be baked into various components.

Until we're ready to migrate that logic into the slice, I don't think we need this state here either.

Reply:
Also (assuming we need this) if we're setting this with a string value, maybe we should assign that string from an object property (like an enum would)?

Reply:
its set to basic after the user has used the gas modal to set a custom gas price. When this happens the advanced inline gas is shown even on mainnet when our gas estimation is fully functional. A reset button is then shown which sets the mode back to 'basic'. 

I agree with you that this is unclear and that if we want to do this right that bit of logic should be moved into the slice.

Reply:
Made a constant for these values and refined this to be gas **input** status not gas **estimation** status. There are three values:
1. BASIC (basic estimate buttons)
2. INLINE (inline fields)
3. CUSTOM (user has provided custom limit/price from the adv gas modal) 

2 is set whenever the gas estimation slice fails or gets its value from eth_gasPrice. in the selector we check if we're on mainnet/IN_TEST and will override to inline if we're on non mainnet and not IN_TEST. this allows me to remove the logic from send-gas-row that was piecing together all of these values as @Gudahtt mentioned.

Reply:
Makes sense! I must have misunderstood the `onSubmit` function I was looking at. 

I'm still not entirely sure that we need to have INLINE and BASIC state :thinking: Since that's determined by context and by the user's advanced gas preference. But I definitely see the need for CUSTOM as send state - that seems like a necessity. I'll take another look over how INLINE and BASIC are set.

Reply:
It's also set to in-line if they are on any network other than mainnet

Reply:
I think I see now what was throwing me off. It's that the gas mode in the slice isn't the "true" gas mode. The selector gives the real gas mode value, because it overrides the send slice state if appropriate based on the network state and the advanced gas setting.

Maybe this is just a confusing naming/documentation issue then.

Maybe this would be simpler if, in state, we just tracked whether custom gas was set? If it was, then the mode is 'CUSTOM'. If not, then the mode is 'DEFAULT' and the selector uses the network state and advanced gas setting and the gas source data. That would let us get rid of some of the reducer logic, it would be logically equivalent, and it would avoid the "gas mode" in state being mismatched with the "real" gas mode. And it would avoid the partial state duplication dance we have now with the gas source switch.

Reply:
https://github.com/MetaMask/metamask-extension/pull/10965/commits/4a85992056167198f9384b13aac0f47f32c5761c -- very cool idea. THis still uses GAS_INPUT_MODES but this is no longer a key in state

Reply:
Looks great! Agreed that `GAS_INPUT_MODES` is still useful - I was thinking the same thing.